core/src/apps/common/signverify.py

```
@@ -27,8 +27,15 @@ def message_digest(coin: CoinInfo, message: bytes) -> bytes:
 
+def is_non_printable(message: str) -> bool:
+    return any(ord(c) < 32 or ord(c) == 127 for c in message)
+
+
 def decode_message(message: bytes) -> str:
     try:
-        return bytes(message).decode()
+        decoded_message = bytes(message).decode()
+        if is_non_printable(decoded_message):
+            return f"0x{hexlify(message).decode()}"
+        return decoded_message
```

Ruff (0.8.2)

Consider moving this statement to an `else` block

(TRY300)